src/main/kotlin/treeLib/bintrees/AVLTree.kt

```diff
@@ -0,0 +1,174 @@
+    }
+
+    private fun rotateLeft(nodeA: AVLNode<K, V>?): AVLNode<K, V>? {
+        if (nodeA != null) {
+            if (nodeA.right != null) {
+                val nodeB = nodeA.right
+                nodeA.right = nodeB?.left
+                nodeB?.left = nodeA
+                fixHeight(nodeA)
+                fixHeight(nodeB)
+                return nodeB
+            } else {
+                throw NullPointerException("Right child node cannot be null.")
+            }
+        } else {
+            throw NullPointerException("Node cannot be null.")
+        }
+    }
+
+    private fun rotateRight(nodeA: AVLNode<K, V>?): AVLNode<K, V>? {
+        if (nodeA != null) {
+            if (nodeA.left != null) {
+                val nodeB = nodeA.left
+                nodeA.left = nodeB?.right
+                nodeB?.right = nodeA
+                fixHeight(nodeA)
+                fixHeight(nodeB)
+                return nodeB
+            } else {
+                throw NullPointerException("Left child node cannot be null.")
+            }
+        } else {
+            throw NullPointerException("Node cannot be null.")
+        }
+    }
+
+    private fun height(node: AVLNode<K, V>?): Int {
+        if (node == null) {
+            return 0
+        }
+        return node.height
+    }
+
+    private fun balanceFactor(node: AVLNode<K, V>?): Int {
+        if (node == null) {
+            throw NullPointerException("Node cannot be null.")
+        }
+        return height(node.right) - height(node.left)
+    }
+
+    private fun fixHeight(node: AVLNode<K, V>?) {
+        if (node != null) {
+            val heightLeft = height(node.left)
+            val heightRight = height(node.right)
+            if (heightLeft > heightRight) {
+                node.height = heightLeft + 1
+            } else {
+                node.height = heightRight + 1
+            }
+        }
+    }
+
+    private fun balanceNode(node: AVLNode<K, V>?): AVLNode<K, V>? {
+        fixHeight(node)
+        if (balanceFactor(node) == 2) {
+            if (node != null) {
+                if (balanceFactor(node.right) < 0) {
+                    node.right = rotateRight(node.right)
+                }
+                if (node == root) {
+                    root = rotateLeft(node)
+                    return root
+                }
+                return rotateLeft(node)
+            }
+        }
+        if (balanceFactor(node) == -2) {
+            if (node != null) {
+                if (balanceFactor(node.left) > 0) {
+                    node.left = rotateLeft(node.left)
+                }
+                if (node == root) {
+                    root = rotateRight(node)
+                    return root
+                }
+                return rotateRight(node)
+            }
+        }
+        return node
+    }
+
+    override fun height(): Int? {
+        return if (root == null) {
+            0
+        } else {
+            root?.height
+        }
```

Comment:
Здесь вообще ошибок нет. Я могу только посоветовать использовать ?: . Ведь смысл один и тот же, а писать на три строки меньше. Примерно так:
```suggestion
       return root?.height ?: 0
```

Reply:
Спасибо за предложение. Думала об использовании ?:, но показалось, будто более длинный вариант более понятен, на случай если читающий код не очень знаком с тернарными операторами (и в частности с ?:).